File: uisdk/src/test/java/com/karhoo/uisdk/screen/booking/map/BookingMapPresenterTest.java
package com.karhoo.uisdk.screen.booking.map;

// TODO: Create kotlin file for this.

//@RunWith(MockitoJUnitRunner.class)
//public class BookingMapPresenterTest {

//    @Mock
//    BookingMapMVP.View view;
//    @Mock
//    BookingMapStategy.Presenter pickupPresenter;
//    @Mock
//    BookingMapStategy.Presenter pickupDropOffPresenter;
//    @Mock
//    LiveBookingStatus bookingStatus;
//    @Mock
//    com.karhoo.sdk.api.model.LatLng kLatLong;
//    @Mock
//    LocationDetails locationDetails;
//    @Mock
//    CompositeDisposableContainer compositeDisposable;
//    @Mock
//    Analytics analytics;
//
//    @Captor
//    ArgumentCaptor<BookingMapStategy.Owner> captor;
//
//    BookingMapPresenter presenter;
//    LatLng latLng;
//
//    @Before
//    public void setUp() throws Exception {
//        presenter = new BookingMapPresenter(view, pickupPresenter,
//                pickupDropOffPresenter, compositeDisposable, analytics);
//    }
//
//    /**
//     * Given:   The booking status is pickup only
//     * When:    Looking at the map
//     * Then:    The active presenter should be the pickupOnly presenter
//     */
//    @Test
//    public void presenterPickupShouldBeSetForTheCurrentBookingStatus() throws Exception {
//        latLng = new LatLng(0, 0);
//        when(locationDetails.getLatLng()).thenReturn(kLatLong);
//        when(kLatLong.getLongitude()).thenReturn(0D);
//        when(kLatLong.getLatitude()).thenReturn(0D);
//        LiveBookingStatus bookingStatus = new LiveBookingStatus();
//        bookingStatus.setPickup(locationDetails);
//        bookingStatus.setDestination(null);
//        ObservableVariable<LiveBookingStatus> observable = new ObservableVariable<>();
//        presenter.observeBookingStatus(observable);
//        observable.setValue(bookingStatus);
//        presenter.mapMoved(latLng);
//
//        verify(pickupPresenter, atLeastOnce()).mapMoved(latLng);
//        verify(pickupDropOffPresenter, never()).mapMoved(latLng);
//    }
//
//    /**
//     * Given:   The booking status is pickup only
//     * When:    Looking at the map
//     * Then:    The active presenter should be the pickupOnly presenter
//     */
//    @Test
//    public void presenterPickupDropoffShouldBeSetForTheCurrentBookingStatus() throws Exception {
//        latLng = new LatLng(0, 0);
//        LiveBookingStatus bookingStatus = new LiveBookingStatus();
//        bookingStatus.setPickup(locationDetails);
//        bookingStatus.setDestination(locationDetails);
//        ObservableVariable<LiveBookingStatus> observable = new ObservableVariable<>();
//        presenter.observeBookingStatus(observable);
//        observable.setValue(bookingStatus);
//        presenter.mapMoved(latLng);
//
//        verify(pickupDropOffPresenter, atLeastOnce()).mapMoved(latLng);
//        verify(pickupPresenter, never()).mapMoved(latLng);
//    }
//
//    /**
//     * Given:   The user clears the destination
//     * When:    The map reverts back to the pickup location
//     * Then:    The map should be asked to clear all the markers
//     */
//    @Test
//    public void markersGetClearedWhenClearingDestination() throws Exception {
//        when(locationDetails.getLatLng()).thenReturn(kLatLong);
//        when(kLatLong.getLongitude()).thenReturn(0D);
//        when(kLatLong.getLatitude()).thenReturn(0D);
//        LiveBookingStatus bookingStatus = new LiveBookingStatus();
//        bookingStatus.setPickup(locationDetails);
//        bookingStatus.setDestination(locationDetails);
//        ObservableVariable<LiveBookingStatus> observable = new ObservableVariable<>();
//        presenter.observeBookingStatus(observable);
//        observable.setValue(bookingStatus);
//        bookingStatus.setDestination(null);
//        observable.setValue(bookingStatus);
//
//        verify(view, atLeastOnce()).clearMarkers();
//
//    }
//
//    /**
//     * Given:   A user has pressed locate meh
//     * When:    That requested is passed on to the main presenter
//     * Then:    The presenter should forward that to the correct place
//     */
//    @Test
//    public void locateMehGoesToTheCorrectPresenter() throws Exception {
//        when(locationDetails.getLatLng()).thenReturn(kLatLong);
//        when(kLatLong.getLongitude()).thenReturn(0D);
//        when(kLatLong.getLatitude()).thenReturn(0D);
//        LiveBookingStatus bookingStatus = new LiveBookingStatus();
//        bookingStatus.setPickup(locationDetails);
//        bookingStatus.setDestination(null);
//        ObservableVariable<LiveBookingStatus> observable = new ObservableVariable<>();
//        presenter.observeBookingStatus(observable);
//        observable.setValue(bookingStatus);
//        presenter.locateUserPressed(new LatLng(0D, 0D));
//
//        verify(pickupPresenter, atLeastOnce()).locateUserPressed((LatLng) any());
//    }
//
//    /**
//     * Given:   A pickup and drop off is set
//     * When:    The map comes into view
//     * Then:    The map should be asked to zoom to the markers
//     */
//    @Test
//    public void mapZoomiesToMarkersWhenPickupAndDropoffAreSet() throws Exception {
//        when(locationDetails.getLatLng()).thenReturn(kLatLong);
//        when(kLatLong.getLongitude()).thenReturn(0D);
//        when(kLatLong.getLatitude()).thenReturn(0D);
//        doAnswer(new Answer() {
//            @Override
//            public Object answer(InvocationOnMock invocation) throws Throwable {
//                captor.getValue().zoomMapToMarkers();
//                return null;
//            }
//        }).when(pickupDropOffPresenter).locateUserPressed((LatLng) any());
//        LiveBookingStatus bookingStatus = new LiveBookingStatus();
//        bookingStatus.setPickup(locationDetails);
//        bookingStatus.setDestination(locationDetails);
//        ObservableVariable<LiveBookingStatus> observable = new ObservableVariable<>();
//        verify(pickupDropOffPresenter, atLeastOnce()).setOwner(captor.capture());
//        presenter.observeBookingStatus(observable);
//        observable.setValue(bookingStatus);
//        presenter.locateUserPressed(latLng);
//
//        verify(view, atLeastOnce()).addMarkers((com.karhoo.sdk.api.model.LatLng) any(), (com.karhoo.sdk.api.model.LatLng) any());
//        verify(pickupDropOffPresenter, atLeastOnce()).locateUserPressed((LatLng) any());
//    }
//
//    /**
//     * Given:   When a user drags the map
//     * When:    there is no destination set
//     * Then:    The appropriate presenter should be called
//     */
//    @Test
//    public void userDraggingMapGetsPassedOntoPresenter() throws Exception {
//        when(locationDetails.getLatLng()).thenReturn(kLatLong);
//        when(kLatLong.getLongitude()).thenReturn(0D);
//        when(kLatLong.getLatitude()).thenReturn(0D);
//        LiveBookingStatus bookingStatus = new LiveBookingStatus();
//        bookingStatus.setPickup(locationDetails);
//        bookingStatus.setDestination(null);
//        ObservableVariable<LiveBookingStatus> observable = new ObservableVariable<>();
//        presenter.observeBookingStatus(observable);
//        observable.setValue(bookingStatus);
//        presenter.zoom((LatLng) any());
//
//        verify(view, atLeastOnce()).zoom((LatLng) any());
//    }
//
//    /**
//     * Given:   A user moves the map
//     * When:    The event is captured
//     * Then:    Its passed to the correct presenter
//     */
//    @Test
//    public void whenMapIsMovedThePresenterIsAlerted() throws Exception {
//        when(locationDetails.getLatLng()).thenReturn(kLatLong);
//        when(kLatLong.getLongitude()).thenReturn(0D);
//        when(kLatLong.getLatitude()).thenReturn(0D);
//        LiveBookingStatus bookingStatus = new LiveBookingStatus();
//        bookingStatus.setPickup(locationDetails);
//        bookingStatus.setDestination(null);
//        ObservableVariable<LiveBookingStatus> observable = new ObservableVariable<>();
//        presenter.observeBookingStatus(observable);
//        observable.setValue(bookingStatus);
//        presenter.mapDragged();
//
//        verify(pickupPresenter, atLeastOnce()).mapDragged();
//    }
//
//    /**
//     * Given:   I only have a pickup set
//     * When:    I press locate me
//     * Then:    A call should be made to allow to be located
//     */
//    @Test
//    public void whenLocateAndUpdateCallToViewToLocateTheUser() throws Exception {
//        presenter.locateAndUpdate();
//        verify(view, atLeastOnce()).locationUpdatesAllowed();
//    }
//
//    /**
//     * Given:   There is an error
//     * When:    onError called
//     * Then:    error passed through to view
//     */
//    @Test
//    public void onErrorShowErrorInView() {
//        int errorId = R.string.kh_uisdk_booking_error;
//        presenter.onError(errorId);
//        verify(view).showError(errorId);
//    }
//
//    /**
//     * Given:   I only have a pickup set
//     * When:    I press locate me
//     * Then:    An analytic call will fire an event for reverse geo
//     */
//    @Test
//    public void whenLocateAndUpdateAnalyticsWillCallRevGeo() throws Exception {
//        when(locationDetails.getLatLng()).thenReturn(kLatLong);
//        when(kLatLong.getLongitude()).thenReturn(0D);
//        when(kLatLong.getLatitude()).thenReturn(0D);
//        LiveBookingStatus bookingStatus = new LiveBookingStatus();
//        bookingStatus.setPickup(locationDetails);
//        bookingStatus.setDestination(null);
//        ObservableVariable<LiveBookingStatus> observable = new ObservableVariable<>();
//        presenter.observeBookingStatus(observable);
//        observable.setValue(bookingStatus);
//        presenter.locateUserPressed(latLng);
//        verify(analytics).reverseGeo();
//    }
//
//
//    /**
//     * When:    onPause is called
//     * Then:    The CompositeDisposableContainer is disposed
//     */
//    @Test
//    public void compositeDisposableContainerIsDisposedOnStop() throws Exception {
//        presenter.onPause();
//        verify(compositeDisposable).dispose();
//    }

//}

File: uisdk/src/test/java/com/karhoo/uisdk/screen/booking/map/BookingMapPresenterTest.kt
package com.karhoo.uisdk.screen.booking.map

import androidx.lifecycle.LifecycleOwner
import com.google.android.gms.maps.model.LatLng
import com.karhoo.uisdk.R
import com.karhoo.uisdk.analytics.Analytics
import com.karhoo.uisdk.screen.booking.domain.address.BookingStatus
import com.karhoo.uisdk.screen.booking.domain.address.BookingStatusStateViewModel
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.atLeastOnce
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.verify
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.junit.MockitoJUnitRunner
import kotlin.jvm.Throws

@RunWith(MockitoJUnitRunner::class)
class BookingMapPresenterTest {

    internal lateinit var presenter: BookingMapPresenter

    internal var view: BookingMapMVP.View = mock()
    internal var pickupPresenter: BookingMapStategy.Presenter = mock()
    internal var pickupDropOffPresenter: BookingMapStategy.Presenter = mock()
    internal var bookingStatus: BookingStatus = mock()
    internal var analytics: Analytics? = mock()
    internal var lifeCycleOwner: LifecycleOwner = mock()
    internal var bookingStatusStateViewModel: BookingStatusStateViewModel = mock()

    @Before
    fun setUp() {
        presenter = BookingMapPresenter(
                view = view,
                analytics = analytics,
                pickupOnlyPresenter = pickupPresenter,
                pickupDropoffPresenter = pickupDropOffPresenter
                                       )
    }

    /**
     * Given:   The user has given their permission for location
     * When:    A call is made to locationPermissionGranted
     * Then:    The map should reset and the user should be located
     */
    @Test
    @Throws(Exception::class)
    fun `Location permission granted`() {
        presenter.locationPermissionGranted()

        verify(view, atLeastOnce()).resetMap()
        verify(view, atLeastOnce()).locateUser()
    }

    // TODO: These unfinished tests require a refactor of the BookingMapPresenter in relation to
    //  the tests commented out below.
//    /**
//     * Given: The user is loading the app
//     * When: The booking screen appears
//     * Then: Watch booking status
//     */
//    @Test
//    fun `Example Test One`() {
//        val observer = presenter.watchBookingStatus(lifeCycleOwner, bookingStatusStateViewModel)
//    }
//
//    /**
//     * Given: The user is loading the app
//     * When: The booking screen appears
//     * Then: Watch booking status
//     */
//    @Test
//    fun `Example Test One Part Two`() {
//        val observer = presenter.watchBookingStatus(lifeCycleOwner, bookingStatusStateViewModel)
//
//    }

    /**
     * Given: The user is on the booking screen
     * When: The locate me button is pressed
     * Then: Zoom to user location
     */
    @Test
    fun `Zoom to User Location`() {
        val latLng = LatLng(100.0, 200.0)
        presenter.zoom(position = latLng)

        verify(view, atLeastOnce()).zoom(position = latLng)
    }

    /**
     * Given: The user is on the booking screen
     * When: A pickup destination is set
     * Then: Move the map to the marker
     */
    @Test
    fun `Focus map on pickup point`() {
        presenter.zoomMapToMarkers()

        verify(view, atLeastOnce()).zoomMapToOriginAndDestination()
    }

    /**
     * Given: The user is on the booking screen
     * When: An error occurs
     * Then: A snackbar is shown
     */
    @Test
    fun `Show snackbar on error`() {
        val resId: Int = R.string.kh_uisdk_K4001
        presenter.onError(resId, null)

        verify(view).showSnackbar(any())
    }

}
